### PR TITLE
🤙 Introduce useCall hook

### PR DESCRIPTION
**Pull request by HubertNastaly**

### CODE REVIEW

File: packages/example/src/components/TokenList/TokenList.tsx [hunk 1]
```diff
@@ -25,7 +25,15 @@ function useTokensBalance(tokenList?: any[], account?: string | null) {
 export function TokenList() {
   const { account, chainId } = useEthers()
   const { name, logoURI, tokens } = useTokenList(UNISWAP_DEFAULT_TOKEN_LIST_URI, chainId) || {}
-  const balances = useTokensBalance(tokens, account)
+  const { results: balances, error } = useTokensBalance(tokens, account)
+
+  if (error && error.event === 'noNetwork') {
+    return <span>Connect wallet to see tokens list</span>
+  }
+
```
Comment (HubertNastaly) [inline packages/example/src/components/TokenList/TokenList.tsx:33]: Btw is it expected behaviour that when I disconnect all of my wallets from the example website, the `chainId` is still equal `1`?

File: docs/source/guide.rst [hunk 2]
```diff
@@ -138,11 +138,10 @@ The hook will retrieve a balance of an ERC20 token of the provided address.
     address: string | Falsy
   ) {
     const [tokenBalance] =
-      useContractCall(
+      useCall(
         address &&
           tokenAddress && {
-            abi: ERC20Interface, // ABI interface of the called contract
-            address: tokenAddress, // On-chain address of the deployed contract
+            contract: contract: new Contract(tokenAddress, ERC20Interface), // instance of called contract
```
Comment (tt-marek) [inline docs/source/guide.rst:144]: Duplicated: `contract:`

Comment (tt-marek) [reply docs/source/guide.rst:144]: Feature to consider in future PRs:
contract: createContract(tokenAddress, ERC20Interface)

File: packages/core/src/hooks/useCall.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import { ChainCall } from '../providers/chainState/callsReducer'
+import { Contract } from '@ethersproject/contracts'
+
+function warnOnInvalidCall(call: Call | Falsy) {
+  if (!call) {
+    return
+  }
+  const { contract, method, args } = call
+  console.warn(`Invalid contract call: address=${contract.address} method=${method} args=${args}`)
+}
```
Comment (tt-marek) [inline packages/core/src/hooks/useCall.ts:13]: Extract to separate file

File: packages/example/src/components/TokenList/TokenList.tsx [hunk 4]
```diff
@@ -25,7 +25,15 @@ function useTokensBalance(tokenList?: any[], account?: string | null) {
 export function TokenList() {
   const { account, chainId } = useEthers()
   const { name, logoURI, tokens } = useTokenList(UNISWAP_DEFAULT_TOKEN_LIST_URI, chainId) || {}
-  const balances = useTokensBalance(tokens, account)
+  const { results: balances, error } = useTokensBalance(tokens, account)
+
+  if (error && error.event === 'noNetwork') {
```
Comment (tt-marek) [inline packages/example/src/components/TokenList/TokenList.tsx:30]: error should come from `useEthers`

**Comment by noobshow** [comment]
Looking forward to completing and merging this PR

**Comment by noobshow** [comment]
Can't multicall v1 just be phased out since v2 simplifies even the project's goal?

**Comment by noobshow** [comment]
At least, we can phase it out in probably 2 versions from now to not confuse new users

File: docs/source/core.rst [hunk 5]
```diff
@@ -83,37 +83,115 @@ useBlockNumber
 
 Get the current block number. Will update automatically when the new block is mined.
 
-useChainCall
-============
+useCall
+=======
+
+Makes a call to a specific contract and returns the value or an error if present. The hook will cause the component to refresh when a new block is mined and the return value changes.
+A syntax sugar for `useFailableChainCall`_ that uses ABI, function name, and arguments instead of raw data.
```
Comment (tt-marek) [inline docs/source/core.rst:90]: Do we have idea for better name for `useFailableChainCall`?

File: docs/source/core.rst [hunk 6]
```diff
@@ -83,37 +83,115 @@ useBlockNumber
...
+
+**Parameters**
+
+- ``calls: Call | Falsy`` - a single call to a contract , also see `Call`_
+
+**Returns**
+
+- ``{ value: any[] | undefined, error: string | undefined } | undefined``
```
Comment (tt-marek) [inline docs/source/core.rst:98]: Should error be string or more typed?

File: docs/source/core.rst [hunk 7]
```diff
@@ -83,37 +83,115 @@ useBlockNumber
 
 Get the current block number. Will update automatically when the new block is mined.
 
-useChainCall
-============
+useCall
+=======
+
+Makes a call to a specific contract and returns the value or an error if present. The hook will cause the component to refresh when a new block is mined and the return value changes.
```
Comment (tt-marek) [inline docs/source/core.rst:89]: ...specific function of a specific contract...

File: docs/source/core.rst [hunk 8]
```diff
@@ -83,37 +83,115 @@ useBlockNumber
...
+
+**Parameters**
+
+- ``calls: Call | Falsy`` - a single call to a contract , also see `Call`_
+
+**Returns**
+
+- ``{ value: any[] | undefined, error: string | undefined } | undefined``
+
+The hook returns ``CallResult`` type. That is: `undefined` when call didn't return yet or a object ``{ value | error }`` if it did, ``value`` - array of results or undefined if error occured, ``error: string | undefined`` - encountered error or undefined if call was successful.
```
Comment (tt-marek) [inline docs/source/core.rst:100]: Add an example.

File: docs/source/core.rst [hunk 9]
```diff
@@ -83,37 +83,115 @@ useBlockNumber
...
+
+
+useCalls
+================
+Makes calls to specific contracts and returns values or an error if present. The hook will cause the component to refresh when a new block is mined and the return values change.
+A syntax sugar for `useFailableChainCalls`_ that uses ABI, function name, and arguments instead of raw data.
+
+**Parameters**
+
```
Comment (tt-marek) [inline docs/source/core.rst:109]: Add an example.

File: docs/source/core.rst [hunk 10]
```diff
@@ -83,37 +83,115 @@ useBlockNumber
...
-===============
+*Returns*
+
+- ``(string | Falsy)[]`` - encoded results or Falsy values if call didn't return yet or an error occured
+
+
+useFailableChainCall
+=================
+
+Makes a call to a specific contract and returns the value or error if present. The hook will cause the component to refresh whenever a new block is mined and the value is changed.
```
Comment (tt-marek) [inline docs/source/core.rst:155]: A low level function that makes...

File: docs/source/guide.rst [hunk 11]
```diff
@@ -138,11 +138,10 @@ The hook will retrieve a balance of an ERC20 token of the provided address.
     address: string | Falsy
   ) {
     const [tokenBalance] =
-      useContractCall(
```
Comment (tt-marek) [inline docs/source/guide.rst:141]: Needs update.

File: packages/core/src/hooks/useCall.ts [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+          return {
+            value: undefined,
+            error: new utils.Interface(['function Error(string)']).decodeFunctionData(
+              'Error',
+              value
+            )[0] as ErrorMessage, // decode error message,
+          }
+        }
+      }),
+    [results]
```
Comment (tt-marek) [inline packages/core/src/hooks/useCall.ts:49]: Should we us JSON.stringify here?

File: packages/core/src/hooks/useCall.ts [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+          }
+        } else {
+          return {
+            value: undefined,
+            error: new utils.Interface(['function Error(string)']).decodeFunctionData(
+              'Error',
+              value
+            )[0] as ErrorMessage, // decode error message,
+          }
+        }
```
Comment (tt-marek) [inline packages/core/src/hooks/useCall.ts:47]: Extract `decodeCallResult()` with tests!

File: packages/example/src/components/TokenList/TokenList.tsx [hunk 14]
```diff
@@ -1,20 +1,20 @@
...
 import { TextBold } from '../../typography/Text'
 import { TokenIcon } from './TokenIcon'
 import { toHttpPath } from '../../utils'
+import { Contract } from '@ethersproject/contracts'
 
 const UNISWAP_DEFAULT_TOKEN_LIST_URI = 'https://gateway.ipfs.io/ipns/tokens.uniswap.org'
 
 function useTokensBalance(tokenList?: any[], account?: string | null) {
-  return useContractCalls(
+  return useCalls(
```
Comment (tt-marek) [inline packages/example/src/components/TokenList/TokenList.tsx:14]: Please add/update useCall in guides.

Which is better:
```ts
const {value as price1, error as error1} = useCall(pair1, ...) 
const {value as price2, error as error2} = useCall(pair2, ...)
if (error1) {
  alert(error1)
}
```

```ts
const [price1, error1] = useCall(pair1, ...) 
const [price2, error2] = useCall(pair2, ...)
if (error1) {
  alert(error1)
}
```

```ts
const [price1, error1] = useCall(pair1, ...) 
const [price2, error2] = useCall(pair2, ...)
if (error1) {
  alert(price1)
}
```

File: packages/core/src/hooks/useFailableChainCalls.ts [hunk 15]
```diff
@@ -0,0 +1,28 @@
+import { useEffect, useMemo } from 'react'
+import { MulticallResult, useChainState } from '../providers'
+import { ChainCall } from '../providers'
+import { Falsy } from '../model/types'
+
+export function useFailableChainCalls(calls: (ChainCall | Falsy)[]): MulticallResult[] {
```
Comment (tt-marek) [inline packages/core/src/hooks/useFailableChainCalls.ts:6]: Rename proposals:
`useFailableChainCalls` -> `useRawCall`
`ChainCall` -> `RawCall`

Comment (HubertNastaly) [reply packages/core/src/hooks/useFailableChainCalls.ts]: ✅

File: packages/core/src/hooks/useCall.ts (hunk 12 above)
Comment (HubertNastaly) [reply packages/core/src/hooks/useCall.ts]: After a longer while, I think it's not necessary as there will be a new instance of results created when anything changes in underlying part.

File: packages/core/src/hooks/useCall.ts (hunk 13 above)
Comment (HubertNastaly) [reply packages/core/src/hooks/useCall.ts]: ✅

File: docs/source/core.rst (hunk 10 above)
Comment (HubertNastaly) [reply docs/source/core.rst]: ✅

File: docs/source/core.rst (hunk 7 above)
Comment (HubertNastaly) [reply docs/source/core.rst]: ✅

File: docs/source/core.rst (hunk 6 above)
Comment (HubertNastaly) [reply docs/source/core.rst]: ✅

File: docs/source/guide.rst (hunk 11 above)
Comment (HubertNastaly) [reply docs/source/guide.rst:141]: ✅

File: docs/source/guide.rst [hunk 16]
```diff
@@ -160,12 +159,11 @@ Another example is useTokenAllowance hook. Instead of balanceOf, we use allowanc
     spenderAddress: string | Falsy
   ) {
     const [allowance] =
-      useContractCall(
+      useCall(
```
[the remark this answers is not in the document]
Comment (HubertNastaly) [reply docs/source/guide.rst:166]: ✅

File: docs/source/core.rst (hunk 9 above)
Comment (HubertNastaly) [reply docs/source/core.rst:126]: ✅

File: docs/source/core.rst [hunk 17]
```diff
@@ -83,37 +83,150 @@ useBlockNumber
...
+
+*Parameters*
+
+- ``call: ChainCall | Falsy`` - a single call, also see `ChainCall`_. A call can be `Falsy`, as it is important to keep the same ordering of hooks even if in a given render cycle there might be not enough information to perform a call.
+
+*Returns*
+
+- ``{ success: boolean, value: string } | undefined`` - result of multicall call
+
+The hook returns ``RawCallResult`` type. That is: ``undefined`` when call didn't return yet or object ``{ success: boolean, value: string }`` if it did, ``success`` - boolean indicating whtether call was successful or not, ``value`` - encoded result when success is ``true`` or encoded error message when success is ``false``
```
Comment (HubertNastaly) [inline docs/source/core.rst:203]:
```suggestion
The hook returns ``RawCallResult`` type. That is: ``undefined`` when call didn't return yet or object ``{ success: boolean, value: string }`` if it did, ``success`` - boolean indicating whtether call was successful or not, ``value`` - encoded result when success is ``true`` or encoded error message when success is ``false``.
```

File: docs/source/guide.rst [hunk 18]
```diff
@@ -186,7 +192,7 @@ Documentation for hooks is available :ref:`here <core:Hooks>`.
 Using hooks considerations
 ==========================
 
-There are some important considerations when using hooks based on `useChainCall`, `useChainCalls` and `useContractCalls`.
+There are some important considerations when using hooks based on `useFailableChainCall`, `useFailableChainCalls`, `useCall` and `useCalls`.
```
Comment (HubertNastaly) [inline docs/source/guide.rst:195]:
```suggestion
There are some important considerations when using hooks based on `useRawCall`, `useRawCalls`, `useCall` and `useCalls`.
```

File: docs/source/core.rst (hunk 17 above)
Comment (HubertNastaly) [inline docs/source/core.rst:203]:
```suggestion
The hook returns ``RawCallResult`` type. That is: ``undefined`` when call didn't return yet or object ``{ success: boolean, value: string }`` if it did, ``success`` - boolean indicating whether call was successful or not, ``value`` - encoded result when success is ``true`` or encoded error message when success is ``false``
```

File: docs/source/core.rst [hunk 19]
```diff
@@ -83,37 +83,150 @@ useBlockNumber
...
+
+*Parameters*
+
+- ``call: ChainCall | Falsy`` - a single call, also see `ChainCall`_. A call can be `Falsy`, as it is important to keep the same ordering of hooks even if in a given render cycle there might be not enough information to perform a call.
+
+*Returns*
+
+- ``{ success: boolean, value: string } | undefined`` - result of multicall call
+
+The hook returns ``RawCallResult`` type. That is: ``undefined`` when call didn't return yet or object ``{ success: boolean, value: string }`` if it did, ``success`` - boolean indicating whtether call was successful or not, ``value`` - encoded result when success is ``true`` or encoded error message when success is ``false``.
```
Comment (HubertNastaly) [inline docs/source/core.rst:203]:
```suggestion
The hook returns ``RawCallResult`` type. That is: ``undefined`` when call didn't return yet or object ``{ success: boolean, value: string }`` if it did, ``success`` - boolean indicating whether call was successful or not, ``value`` - encoded result when success is ``true`` or encoded error message when success is ``false``.
```

File: docs/source/core.rst (hunk 8 above)
Comment (HubertNastaly) [reply docs/source/core.rst]: ✅

File: packages/extension/test/providers/events/reducer.test.ts [hunk 20]
```diff
@@ -351,7 +355,7 @@ describe('reducer', () => {
           multicallAddress: MULTICALL,
           state: {
             [ADDRESS_A]: {
-              '0xdead': '0xbeef',
+              '0xdead': createSuccessResult('0xbeef'),
```
Comment (tt-marek) [inline packages/extension/test/providers/events/reducer.test.ts:358]: inline?

File: packages/example/src/index.tsx [hunk 21]
```diff
@@ -8,6 +8,7 @@ const config = {
   readOnlyUrls: {
     [Mainnet.chainId]: 'https://mainnet.infura.io/v3/14a0951f47e646c1b241aa533e150219',
   },
+  multicallVersion: 2 as const,
```
Comment (tt-marek) [inline packages/example/src/index.tsx:11]: Should example work with non-multicall-2 enabled networks?

File: packages/core/test/multicall2.test.ts [hunk 22]
```diff
@@ -50,39 +52,87 @@ describe('Multicall2', () => {
...
+      {
+        address: tokenContract.address,
+        data: erc20Interface.encodeFunctionData('balanceOf', [tokenContract.address]),
+      },
+    ]
+
+    const blockNumber = await mockProvider.getBlockNumber()
+    const result = await multicall2(mockProvider, multicallContract.address, blockNumber, calls)
+
+    let { value, success } = result[calls[0].address]![calls[0].data] || {}
```
Comment (tt-marek) [inline packages/core/test/multicall2.test.ts:89]: `{value, success} = getResultForCall(call[0])`

File: packages/core/src/hooks/useCall.ts [hunk 23]
```diff
@@ -0,0 +1,20 @@
...
+  args: any[]
+}
+
+export function useCall(call: Call | Falsy): CallResult {
+  return useCalls([call])[0]
+}
+
+export function useCalls(calls: (Call | Falsy)[]): CallResult[] {
+  const results = useRawCalls(calls.map(encodeCallData))
+  return useMemo(() => results.map((result, idx) => decodeCallResult(calls[idx], result)), [results])
```
Comment (tt-marek) [inline packages/core/src/hooks/useCall.ts:19]: Use `JSON.stringify(results)`?

File: packages/core/src/helpers/calls.ts [hunk 24]
```diff
@@ -0,0 +1,65 @@
...
+      value: call.contract.interface.decodeFunctionResult(call.method, value) as any[],
+      error: undefined,
+    }
+  } else {
+    const errorMessage: string = new utils.Interface(['function Error(string)']).decodeFunctionData('Error', value)[0]
+    return {
+      value: undefined,
+      error: new CallError(errorMessage),
+    }
+  }
```
Comment (tt-marek) [inline packages/core/src/helpers/calls.ts:64]: Can decodeFunctionResult throw an exception? Should we handle that separately?

**Review by tt-marek — APPROVED** [review]

**Pull request merged**